src/composer.rs

```diff
@@ -0,0 +1,1052 @@
+    const IDENTITY: WitnessPoint = WitnessPoint::new(Self::ZERO, Self::ONE);
+
+    /// Create an empty constraint system.
+    ///
+    /// This shouldn't be used directly; instead, use [`Self::initialized`]
+    #[deprecated(
+        since = "13.0",
+        note = "this function is meant for internal use. call `initialized` instead"
+    )]
+    fn uninitialized(capacity: usize) -> Self;
+
+    /// Constraints count
+    fn constraints(&self) -> usize;
+
+    /// Allocate a witness value into the composer and return its index.
+    #[deprecated(
+        since = "13.0",
+        note = "this function is meant for internal use. call `append_witness` instead"
+    )]
+    fn append_witness_internal(&mut self, witness: BlsScalar) -> Witness;
+
+    /// Append a new width-4 poly gate/constraint.
+    #[deprecated(
+        since = "13.0",
+        note = "this function is meant for internal use. call `append_custom_gate` instead"
+    )]
+    fn append_custom_gate_internal(&mut self, constraint: Constraint);
+
+    /// PLONK runtime controller
+    fn runtime(&mut self) -> &mut Runtime;
+
+    /// Allocate a witness value into the composer and return its index.
+    fn append_witness<W: Into<BlsScalar>>(&mut self, witness: W) -> Witness {
+        let witness = witness.into();
+
+        #[allow(deprecated)]
+        let witness = self.append_witness_internal(witness);
+
+        let v = self[witness];
+        self.runtime()
+            .event(RuntimeEvent::WitnessAppended { w: witness, v });
+
+        witness
+    }
+
+    /// Append a new width-4 poly gate/constraint.
+    fn append_custom_gate(&mut self, constraint: Constraint) {
+        self.runtime()
+            .event(RuntimeEvent::ConstraintAppended { c: constraint });
+
+        #[allow(deprecated)]
+        self.append_custom_gate_internal(constraint)
+    }
+
+    /// Performs a logical AND or XOR op between the inputs provided for the
+    /// specified number of bits (counting from the least significant bit).
+    ///
+    /// Each logic gate adds `(num_bits / 2) + 1` gates to the circuit to
+    /// perform the whole operation.
+    ///
+    /// ## Constraint
+    /// - is_component_xor = 1 -> Performs XOR between the first `num_bits` for
+    ///   `a` and `b`.
+    /// - is_component_xor = 0 -> Performs AND between the first `num_bits` for
+    ///   `a` and `b`.
+    ///
+    /// # Panics
+    /// This function will panic if the num_bits specified is not even, ie.
+    /// `num_bits % 2 != 0`.
+    fn append_logic_component(
+        &mut self,
+        a: Witness,
+        b: Witness,
+        num_bits: usize,
+        is_component_xor: bool,
+    ) -> Witness {
+        // the bits are iterated as chunks of two; hence, we require an even
+        // number
+        debug_assert_eq!(num_bits & 1, 0);
+
+        let num_bits = cmp::min(num_bits, 256);
+        let num_quads = num_bits >> 1;
+
+        let bls_four = BlsScalar::from(4u64);
```

Comment:
I thought about that, but then it will be an unreadable number. Its not just `BlsScalar([4, 0, 0,0])` because we also multiply by `R2`. The compiler would optimize this to constant anyway. But I agree, its better to have a constant. Added!